cpp_test_suite/cpp_test_ds/DevTest.cpp

```diff
@@ -2612,8 +2612,16 @@ void DevTest::cmd_push_state_status_event()
 	set_change_event("State",true,false);
 	set_change_event("Status",true,false);
 
-    this->push_change_event("state");
-    this->push_change_event("status");
+	struct timeval tv;
+	gettimeofday(&tv, nullptr);
+
+	Tango::DevState state = Tango::DevState::MOVING;
+	Tango::DevString status = "some status";
+
+	//set_state(state);
+	//set_status(status);
```

Comment:
Currently it is required to manually set state and/or status before pushing them, see commented lines. Otherwise these values are only send to client and not set in the device i.e. in a few seconds (depending on server side polling rate) client will get an event with an old value:

```
I have received an event for attribute tango://172.17.0.3:10000/test/debian8/10/Status
Quality = 0
Wed Dec 12 16:42:25 2018 (1544629345,263640 sec) : Status (dim_x = 1, dim_y = 0, w_dim_x = 0, w_dim_y = 0, Data quality factor = VALID, Data format = SCALAR, Data type = DevString)
Element number [0] = some status

I have received an event for attribute tango://172.17.0.3:10000/test/debian8/10/Status
Quality = 0
Wed Dec 12 16:42:34 2018 (1544629354,286715 sec) : Status (dim_x = 1, dim_y = 0, w_dim_x = 0, w_dim_y = 0, Data quality factor = VALID, Data format = SCALAR, Data type = DevString)
Element number [0] = The device is in ON state.
```

Is this OK? 

From API point of view I would add a couple of methods:

```c++
push_state_change_event(DevState,...)
//and
push_status_change_event(DevString,...)
```

Reply:
> Currently it is required to manually set state and/or status before pushing them, see commented lines. Otherwise these values are only send to client and not set in the device i.e. in a few seconds (depending on server side polling rate) client will get an event with an old value:
> 
> ```
> I have received an event for attribute tango://172.17.0.3:10000/test/debian8/10/Status
> Quality = 0
> Wed Dec 12 16:42:25 2018 (1544629345,263640 sec) : Status (dim_x = 1, dim_y = 0, w_dim_x = 0, w_dim_y = 0, Data quality factor = VALID, Data format = SCALAR, Data type = DevString)
> Element number [0] = some status
> 
> I have received an event for attribute tango://172.17.0.3:10000/test/debian8/10/Status
> Quality = 0
> Wed Dec 12 16:42:34 2018 (1544629354,286715 sec) : Status (dim_x = 1, dim_y = 0, w_dim_x = 0, w_dim_y = 0, Data quality factor = VALID, Data format = SCALAR, Data type = DevString)
> Element number [0] = The device is in ON state.
> ```
> 
> Is this OK?

Since DeviceImpl::push_change_event(String, DevState *,...) methods are invoking attr.set_value() or attr.set_value_date_quality(), it would be interesting to invoke set_state() or set_status() at some point in these methods when attr_name is "state"  or "status"... or maybe to do it directly in these set_value() or set_value_date_quality() methods when the attribute name is state or status?

> From API point of view I would add a couple of methods:
> 
> ```c++
> push_state_change_event(DevState,...)
> //and
> push_status_change_event(DevString,...)
> ```
Fine with me.
I would even add:
push_status_change_event(const std::string &)